### PR TITLE
switch to go modules

### PR DESCRIPTION
**Pull request by se3000**
Converts the project over to use [go modules](https://github.com/golang/go/wiki/Modules). A couple notes:

- Almost all dependencies work well, the one exception being that we need [gorm to include the new version of the sqlite3 package](https://github.com/jinzhu/gorm/pull/2565). Until then, we have our own [fork of gorm](https://github.com/smartcontractkit/gorm).
- We have a slightly simpler installation process for geth on CI, but it doesn't install the same version of geth that we use for a dependency. Not being locked to the same version is less than ideal, but there were some breaking changes in the API with the minor version bump. The good new is @tylerferrara is already [working on getting us on geth 1.9](https://github.com/smartcontractkit/chainlink/pull/1500).
- Modules do not feel as feature rich as dep, or at least not as well documented. It still feels a bit early, but in general the whole installation process of Chainlink is much easier and we're now in-line with the Golang ecosystem's long term roadmap.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -3,47 +3,44 @@ orbs:
...
       - image: smartcontract/builder:1.0.21
     steps:
       - checkout
       - run: echo $CACHE_VERSION > cache.version
       - restore_cache:
           name: Restore Go Vendor Cache
-          key: v{{ checksum "cache.version" }}-go-vendor-{{ checksum "Gopkg.lock" }}
-      - run: dep ensure -vendor-only
+          key: go-mod-{{ checksum "go.sum" }}
+      - run: go mod tidy
```
Comment (j16r) [inline .circleci/config.yml]: Is this the right command?

**Comment by tylerferrara** [comment]
The Go-Ethereum upgrade is finished & awaits approval #1511

**Comment by j16r** [comment]
Mind rebasing out those `WIP` commits?

**Comment by j16r** [comment]
The usage of `go mod tidy` doesn't make sense to me here, see:

    tidy        add missing and remove unused modules

This would end up changing the lockfile in CI, which seems undesirable. Wouldn't `go mod download` make more sense or `go mod vendor`?

**Comment by se3000** [comment]
Good call. I'd tried replacing it with `go build`, which *should* also download the dependencies, but it didn't work. `go mod download` seems to be the right command.

File: GNUmakefile [hunk 2]
```diff
@@ -31,24 +31,16 @@ TAGGED_REPO := $(REPO):$(DOCKER_TAG)
...
+.PHONY: gomod
+gomod: ## Ensure chainlink's go dependencies are installed.
 	@if [ -z "`which gencodec`" ]; then \
 		go get github.com/smartcontractkit/gencodec; \
 	fi || true
-
-.PHONY: godep
-godep: ## Ensure chainlink's go dependencies are installed.
-	dep ensure -v -vendor-only
+	go mod tidy
```
Comment (j16r) [inline GNUmakefile]: I think this should also be `go mod download`.

File: Dockerfile [hunk 3]
```diff
@@ -12,8 +12,8 @@ COPY GNUmakefile VERSION ./
 COPY tools/bin/ldflags ./tools/bin/
 
 # Do dep ensure in a cacheable step
-COPY Gopkg.toml Gopkg.lock ./
-RUN make godep
+ADD go.* ./
+RUN go mod tidy
```
Comment (j16r) [inline Dockerfile]: And this.

File: tools/ci/Dockerfile [hunk 4]
```diff
@@ -6,8 +6,8 @@ WORKDIR /go/src/github.com/smartcontractkit/chainlink
 
 # Do dependency installs first, since these will change less than the full
 # source tree and can get cached
-ADD Gopkg.* ./
-RUN dep ensure -vendor-only
+ADD go.* ./
+RUN go mod tidy
```
Comment (j16r) [inline tools/ci/Dockerfile]: This too.

File: tools/ci/appveyor.yml [hunk 5]
```diff
@@ -27,9 +27,7 @@ install:
   - gcc --version
   - go version
   - go env
-  - go get -u -v github.com/golang/dep/cmd/dep
-  # Install Go dependencies.
-  - dep ensure
+  - go mod tidy
```
Comment (j16r) [inline tools/ci/appveyor.yml]: Also this one.

**Pull request merged**